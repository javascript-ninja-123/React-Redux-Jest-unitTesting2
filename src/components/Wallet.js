import React,{Component} from 'react';
import {connect} from 'react-redux';
import {depositBalance,withDrawBalance,resetBalance} from '../actions';
import { Input,Button } from 'semantic-ui-react'

export class WALLET extends Component {
    constructor(props){
    	super(props);
    	this.state = {balance:''};
      this.renderWithDrawlButton = this.renderWithDrawlButton.bind(this);
      this.disabledButton = this.disabledButton.bind(this)
    }
    onChange(e){
      typeof e.target.value === 'number' ?
      this.setState({balance:parseInt(e.target.value,10)})
      :
      this.setState({balance:e.target.value})
    }
    onClickDeposit(){
      this.props.depositBalance(parseInt(this.state.balance,10))
      this.setState({balance:''})
    }
    onClickWithdrawal(){
      this.props.withDrawBalance(parseInt(this.state.balance,10))
        this.setState({balance:''})
    }
    onClickReset(){
      this.props.resetBalance(0)
    }
    renderWithDrawlButton(){
    return  this.props.balance > 0
      ?
      <div>
      <Button
        className='withdraw-btn'
        onClick={this.onClickWithdrawal.bind(this)}
        >
          Withdrawl
        </Button>
        <Button
         className='reset-btn'
         onClick={this.onClickReset.bind(this)}
         >
         Reset
         </Button>
         </div>
      : ''
    }
    disabledButton(){
        return this.state.balance === ''
        ?
        <Button
        className='deposit-btn'
        onClick={this.onClickDeposit.bind(this)}
        disabled={true}
        >
          Deposit
        </Button>
        :
        <Button
        className='deposit-btn'
        disabled={false}
        onClick={this.onClickDeposit.bind(this)}
        >
          Deposit
        </Button>
    }
    render() {
        return (
            <div className="WALLET">
              <h3 className='balance'>Wallet ballance {this.props.balance}</h3>
               <Input placeholder='$ dollars'
               className='balance-input'
               value={this.state.balance}
               type='text'
               onChange={this.onChange.bind(this)}
               />
               {this.disabledButton()}
              {this.renderWithDrawlButton()}
            </div>
        );
    }
}

function mapStateToProps({balance}) {
  return{balance}
}

export default connect(mapStateToProps,{depositBalance,withDrawBalance,resetBalance})(WALLET)
