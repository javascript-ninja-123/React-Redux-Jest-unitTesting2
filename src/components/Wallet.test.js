import React from 'react';
import {shallow} from 'enzyme';
import setupTest from '../setupTest';
import {WALLET} from './Wallet';

describe('Wallet', () => {
    const mockDeposit = jest.fn();
    const mockWithdraw = jest.fn();
    const mockReset = jest.fn();
    const props = {balance: 20,depositBalance:mockDeposit,withDrawBalance:mockWithdraw,resetBalance:mockReset};
    const wallet = shallow(<WALLET {...props}/>)

    describe('HTML (static)', () => {
      it('renders it correctly', () => {
        expect(wallet).toMatchSnapshot();
      })
      it('renders a input to deposit or withdraw from the balance', () => {
        expect(wallet.find('.balance-input').exists()).toBe(true)
      })
      it('contains a state balance from the begining', () => {
        expect(wallet.state().balance).toEqual('')
      })
      it('does not render a withdrawl button when balance is 0', () => {
        const props = {depositBalance:mockDeposit,withDrawBalance:mockWithdraw};
        const wallet = shallow(<WALLET {...props}/>)
        expect(wallet.find('.withdraw-btn').exists()).toBe(false)
      })
      it('render a withdrawal button when balance is bigger than 0', () => {
        const props = {balance:30,depositBalance:mockDeposit,withDrawBalance:mockWithdraw};
        const wallet = shallow(<WALLET {...props}/>)
        expect(wallet.find('.withdraw-btn').exists()).toBe(true)
      })
      it('renders a reset button', () => {
        expect(wallet.find('.reset-btn').exists()).toBe(true);
      })
    })
    describe('Javascript (behavior)',() => {
      const dollar = 100;
      beforeEach(() => {
        wallet.find('.balance-input').simulate('change',{target:{value:dollar}})
      })
      it('change the balance input', () => {
        expect(wallet.state().balance).toEqual(parseInt(dollar,10))
      })
      it('shows the right money',() => {
        expect(wallet.find('.balance').text()).toEqual('Wallet ballance 20')
      })
      describe('user wants to make deposit', () => {
        beforeEach(() => {
          wallet.find('.deposit-btn').simulate('click');
        })
        afterEach(() => {
          wallet.setState({balance:''})
        })
        it('dispatches the deposit() it ecives from props with local banace', () => {
            expect(mockDeposit).toHaveBeenCalledWith(parseInt(dollar,10))
        })
      })
      describe('user wants to withdraw some moneny', () => {
        beforeEach(() => {
          wallet.find('.withdraw-btn').simulate('click')
        })
        afterEach(() => {
          wallet.setState({balance:''})
        })
        it('dispatch the withDrawBalance and it receives from props with local balance', () => {
          expect(mockWithdraw).toHaveBeenCalledWith(parseInt(dollar,10))
        })
      })
      describe('user wants to donate all money', () => {
        beforeEach(() => {
          wallet.find('.reset-btn').simulate('click')
        })
        it('reset everything', () => {
          expect(mockReset).toHaveBeenCalledWith(0)
        })
      })
      describe('disable the deposit button if nothing is typed', () => {
        beforeEach(() => {
          wallet.setState({balance:''})
        })
        it('deposit button is disabled', () => {
          expect(wallet.find('.deposit-btn').prop('disabled')).toEqual(true);
        })
      })
      describe('enable the deposit button if nothing is typed', () => {
        beforeEach(() => {
          wallet.setState({balance:20})
        })
        it('deposit button is enabled', () => {
          expect(wallet.find('.deposit-btn').prop('disabled')).toEqual(false);
        })
      })
    })

})
