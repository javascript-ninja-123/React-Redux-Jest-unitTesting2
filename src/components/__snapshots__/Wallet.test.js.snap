// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet HTML (static) renders it correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WALLET
    balance={20}
    depositBalance={[Function]}
    resetBalance={[Function]}
    withDrawBalance={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h3
          className="balance"
>
          Wallet ballance 
          20
</h3>,
        <Input
          className="balance-input"
          onChange={[Function]}
          placeholder="$ dollars"
          type="text"
          value=""
/>,
        <Button
          as="button"
          className="deposit-btn"
          disabled={true}
          onClick={[Function]}
>
          Deposit
</Button>,
        <div>
          <Button
                    as="button"
                    className="withdraw-btn"
                    onClick={[Function]}
          >
                    Withdrawl
          </Button>
          <Button
                    as="button"
                    className="reset-btn"
                    onClick={[Function]}
          >
                    Reset
          </Button>
</div>,
      ],
      "className": "WALLET",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Wallet ballance ",
            20,
          ],
          "className": "balance",
        },
        "ref": null,
        "rendered": Array [
          "Wallet ballance ",
          20,
        ],
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "class",
        "props": Object {
          "className": "balance-input",
          "onChange": [Function],
          "placeholder": "$ dollars",
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "class",
        "props": Object {
          "as": "button",
          "children": "Deposit",
          "className": "deposit-btn",
          "disabled": true,
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Deposit",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Button
              as="button"
              className="withdraw-btn"
              onClick={[Function]}
>
              Withdrawl
</Button>,
            <Button
              as="button"
              className="reset-btn"
              onClick={[Function]}
>
              Reset
</Button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": null,
            "nodeType": "class",
            "props": Object {
              "as": "button",
              "children": "Withdrawl",
              "className": "withdraw-btn",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Withdrawl",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "class",
            "props": Object {
              "as": "button",
              "children": "Reset",
              "className": "reset-btn",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Reset",
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h3
            className="balance"
>
            Wallet ballance 
            20
</h3>,
          <Input
            className="balance-input"
            onChange={[Function]}
            placeholder="$ dollars"
            type="text"
            value=""
/>,
          <Button
            as="button"
            className="deposit-btn"
            disabled={true}
            onClick={[Function]}
>
            Deposit
</Button>,
          <div>
            <Button
                        as="button"
                        className="withdraw-btn"
                        onClick={[Function]}
            >
                        Withdrawl
            </Button>
            <Button
                        as="button"
                        className="reset-btn"
                        onClick={[Function]}
            >
                        Reset
            </Button>
</div>,
        ],
        "className": "WALLET",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Wallet ballance ",
              20,
            ],
            "className": "balance",
          },
          "ref": null,
          "rendered": Array [
            "Wallet ballance ",
            20,
          ],
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "class",
          "props": Object {
            "className": "balance-input",
            "onChange": [Function],
            "placeholder": "$ dollars",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "class",
          "props": Object {
            "as": "button",
            "children": "Deposit",
            "className": "deposit-btn",
            "disabled": true,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Deposit",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Button
                as="button"
                className="withdraw-btn"
                onClick={[Function]}
>
                Withdrawl
</Button>,
              <Button
                as="button"
                className="reset-btn"
                onClick={[Function]}
>
                Reset
</Button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": null,
              "nodeType": "class",
              "props": Object {
                "as": "button",
                "children": "Withdrawl",
                "className": "withdraw-btn",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Withdrawl",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "class",
              "props": Object {
                "as": "button",
                "children": "Reset",
                "className": "reset-btn",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Reset",
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
